Support TensorList arguments, torch.cat, and narrow in derivatives.yaml

gchanan opened the pull request.

gchanan committed:
Generate torch.cat autograd via ATen.

Most of the change is around supporting generation of:
1) TensorList arguments
2) Arguments to "size", "sizes", i.e. "sizes(dim)"

gchanan commented:
This first two commits here are ATen changes that are equivalent to https://github.com/zdevito/ATen/pull/74 -- I'm just including them here to get valid test output and to show the full issue; we'll actually merge this by first updating ATen and then having the third commit by itself.

Outside of ATen, this basically involves supporting:
1) TensorList arguments in derivatives.yaml
2) sizes(argument) in derivatives.yaml (e.g. collection of sizes of a TensorList along a dimension)
3) size(argument) in derivatives.yaml (e.g. a single dimension size)
4) torch.cat and narrow as double-backwards supporting functions.

gchanan commented:
CC @colesbury

colesbury commented:
the pytorch changes lgtm

apaszke commented on torch/csrc/autograd/function.cpp:
```diff
@@ -30,6 +30,24 @@ auto Function::flags(const variable_list& inputs) -> FunctionFlags {
+  }
+
+  throw std::runtime_error("unexpected");
+  return FunctionFlags();
```
This return shouldn't be necessary

apaszke commented on torch/csrc/autograd/function.cpp:
```diff
@@ -30,6 +30,24 @@ auto Function::flags(const variable_list& inputs) -> FunctionFlags {
   return f;
 }
 
+auto Function::flags(const std::initializer_list<tensor_list> &inputs) -> FunctionFlags {
```
It's better to implement `Function::flags` in terms of begin and end iterators (this can support both `initializer_list`s and `variable_list`s), and add the overloads that dispatch to it in `function.h` so they can be inlined.

apaszke commented on torch/csrc/autograd/function.h:
```diff
@@ -99,6 +99,11 @@ struct Function : std::enable_shared_from_this<Function> {
+  // same as above, but for tensor_list arguments; the use of initializer_list is
+  // to handle overloads correctly, i.e. so { tensor_list } will resolve to this, but
+  // { self, tensor } will resolve to the above.
+  static FunctionFlags flags(const std::initializer_list<tensor_list> &inputs);
```
Wait whaaaaat. Let's just add this overload with begin and end

apaszke commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -123,6 +124,19 @@ inline at::Scalar PythonArgs::scalar(int i) {
+    PyObject* obj = tuple ? PyTuple_GET_ITEM(arg, idx) : PyList_GET_ITEM(arg, idx);
+    res[idx] = reinterpret_cast<THPVariable*>(obj)->cdata;
+  }
+  return res;
```
There's no error checking here, which is worrying for me. It is checked somewhere else, right?

gchanan replied:
I assume you are referring to the variable type checking?  I don't see that being done anywhere else.  I think both this and the existing PythonArgs::tensor above should be calling THPVariable_Check.

colesbury replied:
Type checks are done during parse.

gchanan replied:
Just talked with @colesbury about this; the parse code does enough checking to figure out which signature to dispatch to, but not full checking.  In this particular case, it just matches a TensorList if the python object is a tuple or list, so we actually do need to add the checking as suggested.

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -258,9 +297,31 @@ def get_type(arg_name):
+
+        if arg_sizes_found:
+            option['num_inputs'] = ("+".join(arg_sizes_found) +
+                                    "" if option['num_inputs'] == 0 else " + " + str(option['num_inputs']))
```
This is weird. It seems that 4 lines above `option['num_inputs']` is an int, but here you assign a string to it?

gchanan replied:
it is weird, but it's just dumped to code at the end anyway.  I'll just separate out the integer and string variables.

apaszke replied:
I know. I just can imagine myself doing `if option['num_inputs'] == 5:` in the code and it will silently break because of this

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -331,14 +403,30 @@ def unpack_args():
-            }))
+
+            if arg['type'] == 'TensorList':
+                error_msg = "derivatives don't support specifying both a TensorList and non-TensorList derivative yet"
```
uh let's inline this. it's confusing when it's not in the error path.

I now see that it's used in two places, but I'd still inline it, or I'd put an assertion outside of this for loop

apaszke replied:
BTW what's the problem with having both? they have different templates?

gchanan replied:
Our DERIVATIVE_TENSORLIST template isn't smart enough yet; consider if you had a TensorList "tl" and Tensor "t" argument in that order.  The return value of the tl backward function needs to be assigned to the first tl.size() grad inputs (missing nice python slice syntax).  So instead we just assign the return to the entire grad_inputs variable_list.  Since this is never actually used (at least yet), punting made the most sense.

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -463,25 +560,36 @@ def emit_dispatch(i, option):
         args = []
-        formal_args = ['Tensor & self']
         python_params = args_without_self(option['python_arguments'])
+        has_self = python_params != option['python_arguments']
```
Uh that seems like something that'll easily break in the future, especially that we still haven't clarified what an option even is. Can't we hold a boolean flag in the option?

gchanan replied:
done.

apaszke commented on tools/autograd/gen_variable_type.py:
```diff
@@ -504,12 +612,20 @@ def process_option(name, options):
         dispatch.append('}')
         env['dispatch'] = dispatch
 
+        has_self = 'self' in options[0]['args']
```
Yeah that looks like a better condition

apaszke commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -65,15 +65,36 @@ Tensor & VariableType::checked_unpack(const Tensor & t, const char * name, int p
+     ret[i] = static_cast<VariableImpl*>(t.pImpl)->data;
+   } else {
+   runtime_error("Expected object of type %s but found type %s at position #%d "
+                 "for TensorList argument #%d '%s'",
```
We shouldn't mention TensorList in error messages. Think about the user who has no idea about the C code. It's not helpful

gchanan replied:
The user doesn't know about undefined tensors either, but those are already mentioned in error messages :).

Maybe I'll just say "for iterable argument" so the position number still makes sense.

apaszke replied:
True, however I have no idea if it's possible to even get an undefined tensor in Python (but let's keep the check to be sure). Passing incorrect arguments in sequences is common

apaszke reviewed (commented):
I reviewed only the last commit (PyTorch changes)

apaszke commented:
BTW this is not tested in any way, right?

gchanan commented:
I tested this by commenting out cat in variable.py, but you are correct that it isn't tested via the contiguous integration.

gchanan committed:
Address review comments.

gchanan commented:
@apaszke  ok now that the ATen changes have merged, I forced pushed such that the first commit is equivalent to the old pytorch changes that were reviewed and the next commit addresses the review comments.
